### DISCUSSION TITLE
Tag component considerations

### DISCUSSION THREAD

**Discussion by mbgower** [carbon-design-system/carbon#11215] [Accessibility]
As discussed at CAG, there seem to be at least ~3~ 4 variants of the tag component:
- a read-only indicator, used for classification
- a deletable filter indicator ("x" in corner)
- a selectable tag (logically, if one can delete, one can likely add). The implementation I saw of this on Carbon next involved what was basically a toggle button that a user could put in a pressed or unpressed state
- actionable -- basically a link (see comment)

The Carbon next page also hints at another variant which is almost equivalent of a shopping cart icon, that enumerates the number of tags/filters that are applied.

Challenges and considerations with these variants are addressed in an [accessibility analysis](https://ibm.box.com/s/i60idv4nroqnqlqtm45vbog5cwjwo51e), which includes comments on the existing documentation, recordings of some screen reader issues with a current implementation, and a first take at suggesting how the states for each variant might work.

@laurenmrice and @aagonzales, feel free to bring me in on any discussions on this component.

**Comment by mbgower** [comment]
I saw another actionable tag used in an [ibm.com experimental component](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/canary/web-components-experimental/?path=/story/components-pricing-table--default)

![image](https://user-images.githubusercontent.com/14298245/165552137-57d08741-2863-4a6f-bd4c-07364e56aa81.png)

This demo uses a tag “Merchandising offer” that is an actionable tag (I think it is basically a link?). That breaks the paradigm of tags being informational, not operable (except when used for filtering purposes). It visually is not distinguishable from a ‘read only’, informational tag. That in itself I’d call a design problem.


https://user-images.githubusercontent.com/14298245/165553731-319ea1a3-21c2-48ec-8a3e-00a3e2934cfb.mov

**Reply by mbgower** [reply]
I discussed the above 'button tag' variation with @jwitkowski79's team, who are going to add their designs to this discussion thread as they progress.

**Comment by laurenmrice** [comment]
This is a related issue about an interactive tag https://github.com/carbon-design-system/carbon/issues/7592. We used to have this, but I am not seeing it in our [React storybook](https://react.carbondesignsystem.com/?path=/story/components-tag--default) anymore.

**Comment by laurenmrice** [comment]
The Carbon Platform squad uses selectable tags in a filter menu. Tags that have the selected state will be included in the top filter. Unselected tags will remain visible in the menu but will not be in the top filter. This is a good tag variant to have in our system.

<img width="737" alt="Screen Shot 2022-06-23 at 2 27 56 PM" src="https://user-images.githubusercontent.com/43969356/175369109-f067c42a-f1fb-4297-9a8c-09f0d7bac151.png">
s

**Reply by mbgower** [reply]
Yeah, this one is included in my exploratory. The biggest challenge is the focus indicator on the pushed state. Since it's got a dark outline, a standard focus indicator won't show up. Need an offset --something like a shim of 'white' between the border of the tag and the focus indicator-- or to make it fatter (needs to be at least 2px in WCAG 2.2), etc
<img width="524" alt="image" src="https://user-images.githubusercontent.com/14298245/175390790-d07e98ee-0551-4e47-a247-2a703e67101d.png">

**Comment by mbgower** [comment]
@jwitkowski79 do you have any updates on your exploratory work? Another team (is also getting into this space, and any lessons learned would be helpful.

**Reply by jwitkowski79** [reply]
@mbgower Hello! Unfortunately, our team has not been able to do explorations around this yet. I'll check with Joseph Dombroski to see if I can get the prioritized for our team if it's something we need to tackle sooner than later, and other teams haven't explored solutions to this yet.

**Comment by mbgower** [comment]
Found another use case for tag: in condition statements.
<img width="183" alt="Screenshot 2022-10-27 at 8 26 06 AM" src="https://user-images.githubusercontent.com/14298245/198332848-a7451c62-5eb0-4e46-9953-9a9497dcac3a.png">

Joe Alewine and Alexander Melo were presenting/commenting on today's DSAG on this.
<img width="308" alt="image" src="https://user-images.githubusercontent.com/14298245/198334651-37e936ff-6f89-48cc-9905-82126ddf274f.png">

---

On the same call, I was concerned seeing a column in a table containing deletable tags. This does not seem to be a valid use case (or at least is adding a lot of a complexity to a table that does not have deleting tags as its primary purpose). To me, those should be read-only.
<img width="452" alt="image" src="https://user-images.githubusercontent.com/14298245/198335690-f73e7613-e2a6-4476-851d-d263d106adc8.png">

**Comment by mbgower** [comment]
I realized there is what could be another use case that has come up a number of times on the #accessibility-at-ibm slack channel, including this very meaty thread on [using tag-like styling as a form of content categorization, including in search results](https://ibm-studios.slack.com/archives/C036P1CTN/p1620725449456100)

![image](https://user-images.githubusercontent.com/14298245/198372866-d71afcb0-a930-498d-9c77-aa4f15c84fc7.png)

Note that many [discussions are about colour highlighting of text that do NOT use tags](https://ibm-studios.slack.com/archives/C036P1CTN/p1657037927685059) as part of the solution. Similar considerations apply, so thought it was worthwhile mentioning it, because inevitably someone will use tags this way.
